crates/flowctl/src/source/unbundle.rs

```diff
@@ -0,0 +1,338 @@
+            );
+        }
+    };
+    Ok(Some(new_catalog))
+}
+
+struct Serializer {
+    json: bool,
+}
+
+impl Serializer {
+    pub fn to_vec<T: serde::Serialize>(&self, value: &T) -> anyhow::Result<Vec<u8>> {
+        // Our models embed serde_json's RawValue, which can only be serialized by serde_json.
+        // Do that first.
+        let v = serde_json::to_vec_pretty(value).context("serializing JSON")?;
+        if self.json {
+            Ok(v)
+        } else {
+            // We want YAML, so transcode directly from serialized JSON to YAML.
+            let mut v2 = Vec::new();
+            let mut deser = serde_json::Deserializer::from_slice(&v);
+            serde_transcode::transcode(&mut deser, &mut serde_yaml::Serializer::new(&mut v2))
+                .context("serializing YAML")?;
+
+            Ok(v2)
+        }
+    }
+}
+
+async fn try_write_file(path: &Path, contents: &[u8], existing: Existing) -> anyhow::Result<()> {
+    // Is the file allowed to exist before we create it?
+    let create_new = match existing {
+        Existing::Abort | Existing::Keep => true,
+        _ => false,
+    };
+    let result = fs::OpenOptions::new()
+        .create_new(create_new)
+        .create(true)
+        .write(true)
+        .open(path)
+        .await;
+    match result {
+        Ok(mut file) => {
+            file.write_all(contents).await?;
+            Ok(())
+        }
+        Err(err)
+            if err.kind() == std::io::ErrorKind::AlreadyExists && existing == Existing::Keep =>
+        {
+            tracing::info!(path = %path.display(), "skipping writing file because it already exists and --existing=skip");
+            Ok(())
+        }
+        // This branch also handles the case of `Existing::Error` if the file already exists.
+        Err(err) => Err(anyhow::anyhow!("unable to open file: {}", err)),
+    }
+}
+
+async fn write_file(path: &Path, contents: &[u8], existing: Existing) -> anyhow::Result<()> {
+    match try_write_file(path, contents, existing).await {
+        Ok(()) => {
+            tracing::info!(path = %path.display(), "wrote file");
+            Ok(())
+        }
+        Err(err) => {
+            tracing::error!(path = %path.display(), error = %err, "failed to write file");
+            Err(err).context(format!("failed to write file: '{}'", path.display()))
+        }
+    }
+}
+
+async fn parse_catalog_spec(flow_yaml_path: &Path) -> Result<models::Catalog, anyhow::Error> {
+    let mut existing = fs::File::open(flow_yaml_path).await?;
+    let mut contents = Vec::with_capacity(16 * 1024);
+    existing.read_to_end(&mut contents).await?;
+    let cat = sources::parse_catalog_spec(&contents)?;
+    Ok(cat)
+}
+
+/// Returns the last path component for a given catalog name, which must be a valid catalog name.
+fn base_name(catalog_name: &impl AsRef<str>) -> &str {
+    catalog_name
+        .as_ref()
+        .rsplit_once("/")
+        .expect("catalog names have at least one '/'")
```

Comment:
Is it dangerous to panic here, where everywhere else you're using `anyhow::Result`? i.e will panicking in `base_name` cause some resources to not get cleaned up properly?

Reply:
Good question. I think it's probably safe enough, but it also can't hurt to change it to return a Result. I'll update it.